### PR TITLE
[Model] Add Spark3 Model

### PR DESCRIPTION
**Pull request by KnightYao**
## Summary

This PR adds native support for the Spark 3 model architecture in SGLang.

The implementation introduces the Spark 3 configuration and causal language model, including sliding-window and full attention, head-wise attention output gating, tensor parallelism, pipeline parallelism, and checkpoint weight loading.

## Changes

|File|Description|
|---|---|
|`vllm/model_executor/models/spark3.py`|Implements `Spark3ForCausalLM`, including attention, MLP, decoder layers, TP/PP support, logits computation, and weight loading.|
|`vllm/transformers_utils/configs/spark3.py`|Adds the native `Spark3Config` implementation.|
|`vllm/model_executor/models/registry.py`|Registers `Spark3ForCausalLM` in the vLLM model registry.|
|`vllm/transformers_utils/config.py`|Registers the `spark3` model type and configuration.|
|`vllm/transformers_utils/configs/__init__.py`|Exposes `Spark3Config` through the vLLM config registry.|
|`docs/models/supported_models.md`|Documents Spark3 as a supported text-generation model.|
|`vllm/tool_parsers/spark3_tool_parser.py`|Implements Spark3 XML tool-call parsing, argument conversion, and streaming extraction.|
|`vllm/tool_parsers/__init__.py`|Registers `spark3` in the tool parser registry.|

## Key Technical Details

1. Efficient Architecture：The models adopt an efficiency-oriented architecture with Sliding Window Attention (SWA) and other structural optimizations, improving the balance between model quality, memory usage, and inference efficiency.
2. Native 1M Context：Both the 1.7B and 4B models natively support up to 1 million tokens of context, enabling long-document understanding, repository-level coding, and long-horizon agent tasks.
3. Strong Coding & Agent Capabilities：The models deliver leading coding and agent performance among models of comparable size, with strong capabilities in code generation, tool use, multi-step execution, and long-context reasoning.
4. Faster Inference：Through model architecture and inference-stack optimizations, the models achieve higher inference efficiency and throughput than many comparable-size models, especially in long-context scenarios.
5. Strong K–12 and Gaokao Performance：The models achieve leading K–12 question-answering performance at comparable parameter scales, with particularly strong results on challenging Chinese Gaokao-style tasks.
6. Advanced Post-Training：The release applies a comprehensive post-training recipe based on Scaled Reinforcement Learning and MOPD, improving reasoning, coding, instruction following, agentic capabilities, and alignment.
7. Controllable Thinking Budget：The models support configurable reasoning levels — none, low, medium, and high — allowing applications to dynamically trade off reasoning quality, latency, and inference cost.
8. 200+ Languages：Multilingual coverage has been expanded to more than 200 languages, supporting broad cross-lingual understanding, generation, translation, and localized applications.
    

## Performance

A serving benchmark was performed with **500 concurrent requests**, and all requests completed successfully.

|Metric|Result|
|---|--:|
|Successful requests|500|
|Failed requests|0|
|Benchmark duration|4.70 s|
|Total input tokens|762,489|
|Total generated tokens|63,478|
|Request throughput|106.42 req/s|
|Output token throughput|13,510.08 tok/s|
|Peak output token throughput|16,452.00 tok/s|
|Total token throughput|175,791.36 tok/s|
|Peak concurrent requests|500|

### Latency

|Metric|Mean|Median|P99|
|---|--:|--:|--:|
|TTFT|2122.04 ms|2111.73 ms|4169.35 ms|
|TPOT|11.58 ms|13.39 ms|14.13 ms|
|ITL|12.34 ms|11.13 ms|32.92 ms|

The benchmark completed with a **100% request success rate**. Under a peak concurrency of 500 requests, the server achieved **106.42 requests/s**, approximately **13.5K output tokens/s**, and **175.8K total tokens/s**.

Decode latency remained relatively stable, with a mean TPOT of **11.58 ms** and a P99 TPOT of **14.13 ms**. TTFT was approximately **2.12 s on average** and **4.17 s at P99** under this high-concurrency workload.

### How to Start

Start the OpenAI server with:

```bash
python -m vllm.entrypoints.openai.api_server \
  --model /path/to/spark3-ckpt \
  --trust-remote-code \
  --enable-auto-tool-choice \
  --tool-call-parser spark
```

For Spark3 native tool prompting, `tool_choice` can be set to `required` or
`auto` in the request payload.

### Supported Functionality

- Parse Spark3 XML tool calls into OpenAI `tool_calls`.
- Support `required` and `auto` tool-choice flows for Spark3-native XML output.
- Emit tool-call deltas in streaming responses.
- Convert tool arguments to typed JSON values when schema metadata is present.

### CODE REVIEW

**Review by youkaichao — CHANGES_REQUESTED** [review]
hold for discussions
